src/Cassandra/Metadata.cs

```diff
@@ -133,14 +148,120 @@ public IEnumerable<IPEndPoint> AllReplicas()
+            var currentTicks = DateTime.UtcNow.Ticks;
+            
+            // running this statement synchronously inside the exclusive scheduler deadlocks
+            var ksList = await _schemaParser.GetKeyspaces(retry).ConfigureAwait(false);
+
+            var task = _tokenMapTaskFactory.StartNew(() =>
+            {
+                var lastRebuildTicks = Interlocked.Read(ref _lastRebuildTicks);
+                if (lastRebuildTicks > currentTicks)
+                {
+                    return true;
+                }
+                Interlocked.Exchange(ref _lastRebuildTicks, currentTicks);
+                
+                Metadata.Logger.Info("Retrieving keyspaces metadata");
+                var ksMap = ksList.Select(ks => new KeyValuePair<string, KeyspaceMetadata>(ks.Name, ks));
+                _keyspaces = new ConcurrentDictionary<string, KeyspaceMetadata>(ksMap);
+                Metadata.Logger.Info("Rebuilding token map");
+                if (Partitioner == null)
+                {
+                    throw new DriverInternalError("Partitioner can not be null");
+                }
+                
+                _tokenMap = TokenMap.Build(Partitioner, Hosts.ToCollection(), _keyspaces.Values);
+                return true;
+            });
+
+            return await task.ConfigureAwait(false);
+        }
+
+        internal Task<bool> RemoveKeyspaceFromTokenMap(string name)
+        {
+            var currentTicks = DateTime.UtcNow.Ticks;
+            return _tokenMapTaskFactory.StartNew(
+                () =>
+                {
+                    var lastRebuildTicks = Interlocked.Read(ref _lastRebuildTicks);
+                    if (lastRebuildTicks > currentTicks)
+                    {
+                        return true;
+                    }
+                    
+                    if (_keyspacesLastUpdateTicks.TryGetValue(name, out var lastUpdateTicks) && lastUpdateTicks > currentTicks)
+                    {
+                        return true;
+                    }
+
+                    _keyspacesLastUpdateTicks[name] = currentTicks;
+                    
+                    Metadata.Logger.Verbose("Removing keyspace metadata: " + name);
+                    if (!_keyspaces.TryRemove(name, out var ks))
+                    {
+                        //The keyspace didn't exist
+                        return false;
+                    }
+                    _tokenMap?.RemoveKeyspace(ks);
+                    return true;
+                });
+        }
+
+        internal async Task<KeyspaceMetadata> UpdateTokenMapForKeyspace(string name)
         {
-            Logger.Info("Rebuilding token map");
-            if (Partitioner == null)
+            var currentTicks = DateTime.UtcNow.Ticks;
+            
+            // running this statement synchronously inside the exclusive scheduler deadlocks
+            var keyspaceMetadata = await _schemaParser.GetKeyspace(name).ConfigureAwait(false);
```

Comment:
This can't be put inside a `StartNew()` block as is, as the continuation from the await will be made in a different thread.

I think that if we want to run it in single threaded fashion, it should be something like:

```c#
await factory.StartNew(async () => await _schemaParser.GetKeyspace(name).ConfigureAwait(false)).ConfigureAwait();
// Do checks about data expiration/validity
// ...
// Schedule a new one
await factory.StartNew(() => 
{
  // TODO: call token map
}).ConfigureAwait();

```

Reply:
Yes, I tried putting inside the factory with `.GetAwaiter().GetResult()` so that there's no continuations involved and everything runs on the same thread but unfortunately it deadlocks in some cases. 
Your code example would not guarantee that those two tasks execute sequentially though if I'm understanding it correctly. Something else might get scheduled between those two tasks, right?

Reply:
yes, the sample I've made is no magic bullet.

It would be interesting to try to find ways to implement single threaded async code in the future with nice constructs.

Reply:
My personal choice is [`ActionBlock` of the TPL Dataflow library](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.actionblock-1?view=netcore-2.2) but it won't fix the root issue that is the "recursion" in these methods. 

I do think that I'm going to give it a go and use `ActionBlock` in the near future for some improvements (maybe in the next iteration).